File: 631.design-excel-sum-formula.java
/*
 * @lc app=leetcode id=631 lang=java
 *
 * [631] Design Excel Sum Formula
 */
import java.util.*;

// @lc code=start
class Excel {
    class Cell {
        int v;
        final Map<Cell, Integer> otherCellCnts = new HashMap<>();

        Cell(int _v) { v = _v; };

        // If this cell is defined as the sum formula,
        // store the counts of the cells referenced.
        // T: O(1)
        void add(Cell c) {
            otherCellCnts.put(c, otherCellCnts.getOrDefault(c, 0) + 1);
        }

        // T: O(n)
        int val() {
            if (otherCellCnts.isEmpty()) return v;
            else {
                int sum = 0;
                for (Cell c : otherCellCnts.keySet()) {
                    sum += otherCellCnts.get(c) * c.val();
                }
                return sum;
            }
        }

        // A constant value is set to this cell. Clear the counts.
        // T: O(1)
        void set(int _v) {
            v = _v;
            otherCellCnts.clear();
        }
    }

    private final Cell[][] cells;  // S: O(h * w)

    // T: O(h * w)
    public Excel(int height, char width) {
        final int h = height + 1, w = width - 'A' + 1;
        cells = new Cell[h][w];
        for (int i = 0; i < h; ++i) {
            for (int j = 0; j < w; ++j) {
                cells[i][j] = new Cell(0);
            }
        }
    }
    
    // T: O(1)
    public void set(int row, char column, int val) {
        cells[row][column - 'A'].set(val);
    }

    // T: O(n) if the cell is defined as a sum of other cells
    public int get(int row, char column) {
        return cells[row][column - 'A'].val();
    }
    
    // T: O(n), n := the number of cells to be summed
    public int sum(int row, char column, String[] numbers) {
        Cell c = cells[row][column - 'A'];
        c.set(0);
        for (String s : numbers) {
            int k = s.indexOf(':');
            if (k > 0) {  // range of cells
                int[] start = getPos(s.substring(0, k)),
                      end = getPos(s.substring(k + 1));
                for (int i = start[0]; i <= end[0]; ++i) {
                    for (int j = start[1]; j <= end[1]; ++j) {
                        c.add(cells[i][j]);
                    }
                }
            } else {  // single cell
                int[] pos = getPos(s);
                c.add(cells[pos[0]][pos[1]]);
            }
        }
        return c.val();
    }

    private static int[] getPos(String s) {
        return new int[] {Integer.parseInt(s.substring(1)), s.charAt(0) - 'A'};
    }
}

/**
 * Your Excel object will be instantiated and called as such:
 * Excel obj = new Excel(height, width);
 * obj.set(row,column,val);
 * int param_2 = obj.get(row,column);
 * int param_3 = obj.sum(row,column,numbers);
 */
// @lc code=end

File: 631.design-excel-sum-formula.py
#
# @lc app=leetcode id=631 lang=python3
#
# [631] Design Excel Sum Formula
#
from collections import defaultdict
from typing import *

# @lc code=start
# Hash Map
# S: O(h * w)
OFFSET = ord('A')

class Excel:
    # T: O(h * w)
    def __init__(self, height: int, width: str):
        # each cell: {'v': val, 'sum': list of associated cells {k: (r, c), v: cell freq}}
        self.table: List[List[Dict]] = \
                [[{'v': 0, 'sum': None} for _ in range(ord(width) - OFFSET + 1)] 
                 for _ in range(height)]

    def __pos(self, r: int, c: str) -> Tuple[int, int]:
        r -= 1
        c = ord(c) - OFFSET
        if 0 <= r < len(self.table) and 0 <= c < len(self.table[0]):
            return r, c
        return -1, -1

    # T: O(1)
    def set(self, row: int, column: str, val: int) -> None:
        r, c = self.__pos(row, column)
        if r >= 0 and c >= 0:
            self.table[r][c] = {'v': val, 'sum': None}

    # T: O(h * w)
    def get(self, row: int, column: str) -> int:
        r, c = self.__pos(row, column)
        if r < 0 or c < 0:
            return 0
        cell = self.table[r][c]
        associated_cells = cell.get('sum', None)
        if associated_cells is None:
            return cell['v']
        return sum(self.get(*pos) * freq for pos, freq in associated_cells.items())

    # T: O(h * w)
    def sum(self, row: int, column: str, numbers: List[str]) -> int:
        r, c = self.__pos(row, column)
        if r < 0 or c < 0:
            return 0
        
        associated_cells = defaultdict(int)
        for cell_id in numbers:
            if cell_id == '':
                continue
            parts = cell_id.split(':')
            if len(parts) == 1:
                s_row, s_col = int(parts[0][1:]), parts[0][0]
                associated_cells[(s_row, s_col)] += 1
            elif len(parts) == 2:
                s_row, s_col = int(parts[0][1:]), parts[0][0]
                e_row, e_col = int(parts[1][1:]), parts[1][0]

                for i in range(s_row, e_row + 1):
                    for j in range(ord(s_col) - OFFSET, (ord(e_col) - OFFSET) + 1):
                        associated_cells[(i, chr(j + OFFSET))] += 1

        self.table[r][c]['sum'] = associated_cells
        return self.get(row, column)
        


# Your Excel object will be instantiated and called as such:
# obj = Excel(height, width)
# obj.set(row,column,val)
# param_2 = obj.get(row,column)
# param_3 = obj.sum(row,column,numbers)
# @lc code=end

